File: bms/go/geo/geo_s1_Interval_AddPoint/geo_s1_Interval_AddPoint.go
package test

import (
	"math"
)

func (i geo_s1_Interval) AddPoint(p float64) geo_s1_Interval {
	if math.Abs(p) > math.Pi {
		return i
	}
	if p == -math.Pi {
		p = math.Pi
	}
	if i.fastContains(p) {
		return i
	}
	if i.IsEmpty() {
		return geo_s1_Interval{p, p}
	}
	if geo_s1_positiveDistance(p, i.Lo) < geo_s1_positiveDistance(i.Hi, p) {
		return geo_s1_Interval{p, i.Hi}
	}
	return geo_s1_Interval{i.Lo, p}
}

func (i geo_s1_Interval) fastContains(p float64) bool {
	if i.IsInverted() {
		return (p >= i.Lo || p <= i.Hi) && !i.IsEmpty()
	}
	return p >= i.Lo && p <= i.Hi
}

func (i geo_s1_Interval) IsInverted() bool { return i.Lo > i.Hi }

func (i geo_s1_Interval) IsEmpty() bool { return i.Lo == math.Pi && i.Hi == -math.Pi }

func geo_s1_positiveDistance(a, b float64) float64 {
	d := b - a
	if d >= 0 {
		return d
	}
	return (b + math.Pi) - (a - math.Pi)
}

type geo_s1_Interval struct {
	Lo, Hi float64
}

File: translations_only/success/go/geo/geo_s1_Interval_AddPoint/mistral/BaseRepair/geo_s1_Interval_AddPoint.rs


use std::f64;

const PI: f64 = std::f64::consts::PI;

fn geo_s1_positive_distance(a: f64, b: f64) -> f64 {
    let d = b - a;
    if d >= 0.0 {
        d
    } else {
        (b + PI) - (a - PI)
    }
}

struct GeoS1Interval {
    lo: f64,
    hi: f64,
}

impl GeoS1Interval {
    fn add_point(&self, p: f64) -> GeoS1Interval {
        let mut p = if p.abs() > PI { self.lo } else { p };
        if self.fast_contains(p) {
            return GeoS1Interval { lo: self.lo, hi: self.hi };
        }
        if self.is_empty() {
            return GeoS1Interval { lo: p, hi: p };
        }
        if geo_s1_positive_distance(p, self.lo) < geo_s1_positive_distance(self.hi, p) {
            GeoS1Interval { lo: p, hi: self.hi }
        } else {
            GeoS1Interval { lo: self.lo, hi: p }
        }
    }

    fn fast_contains(&self, p: f64) -> bool {
        if self.is_inverted() {
            return (p >= self.lo || p <= self.hi) && !self.is_empty();
        }
        p >= self.lo && p <= self.hi
    }

    fn is_inverted(&self) -> bool {
        self.lo > self.hi
    }

    fn is_empty(&self) -> bool {
        self.lo == PI && self.hi == -PI
    }
}

